### ISSUE TITLE
Lexical spans

### ISSUE THREAD

**Issue by conartist6** [bablr-lang/bablr-vm#34]
I want to improve the way the VM handles brace pairs. Let's take a look at part of the JS grammar:

```js
export const grammar = {
  productions: productions({
    *Program() {
      while (!(yield match(bnd(sym.EOF)))) {
        yield eat(node`Statement:body`);
      }
    },

    *BlockStatement() {
      const open = yield eat(PN`{`);
      while (yield eatMatch(node`Statement:body`));
      const close = yield eat(PN`}`);
    },
  }),
};
```

Notice that both of these productions have the same basic functionality: consume a statement list, but both work in slightly different ways. The `Program` production keeps checking for EOF before trying to consume an additional statement, and so for a file containing three statements `Program` will only match against the `Statement` production three times. But in `BlockStatement` this are-we-done check doesn't happen, and so a fourth match against the `Statement` production will inevitably occur and fail before `}` is matched.

I would like for both these productions to work the same way, exactly like `Program` the difference will be that the `{` in `BlockStatement` will create a lexical span, such that the lexer will look ahead for `}` and when `}` is found the lexer will pretend that source has ended. This way both productions will be defined to consume statements until the source has ended.

The added advantage is the presence of information about the closing brace at an earlier point in the history, and the guarantee that matching brace pairs will be easiest to implement in the way that allows the engine to make the strongest guarantees about them.

**Comment by conartist6** [comment]
This mechanism of lookahead ending a lexical span will also be crucial to the implementation of expression associativity!

**Comment by conartist6** [comment]
There will need to be something like a `lexicalState~.spans` stack, where `lexicalState.spans[0].guard` is checked repeatedly to ensure that closing braces are recognized and paired with their openers inside the engine. Recognizing a span-closing brace will also pop the spans stack.

**Comment by conartist6** [comment]
What will it look like from the tooling API perspective to establish a span-closing guard?

**Comment by conartist6** [comment]
It should probably be part of the `match` instruction for the opening token, because we want to be sure that it's possible to tell when intercepting a single command whether it's a single token half of a brace pair

**Comment by conartist6** [comment]
Currently we have
```js
yield {
  verb: sym.match,
  value: {
    successEffect: sym.eat,
    failureEffect: sym.fail,
    matchable: {
      grammar: sym.token,
      production: { type: 'Punctuator', value: '{', alterLexicalContext: '...' },
    },
  },
};
```

**Comment by conartist6** [comment]
To the extent that the way this feature is defined impacts how code folding will work in an editor built on the platform I should consider how I want to facilitate folding for things like switch cases, whose bodies are generally thought of as foldable but which do not have well-defined closing tokens as the close of one case is also the beginning of another.

**Comment by conartist6** [comment]
I'm getting a lot of coroutine vibes from the problem space. For example balanced brace matching is clearly the responsibility of the token layer, which should be guaranteeing that valid programs having balanced blocks where the language dictates they should. In a sensible AST the braces already have their own node type such as `BlockStatement` in ESTree.

The problem is that while matching balanced braces may be a token layer responsibility matching what's inside the braces is definitely still going to be the responsibility of the node layer. So my sense is that while right now the node layer only ever calls the token layer a smarter way to build architect the core would be to allow the token and node layers to pass control back and forth as coroutines!

**Comment by conartist6** [comment]
When inside a lexical span the token layer would execute the negative lookahead for the span terminator alongside any token matches made from the node layer. This way the node layer would not be able to accidentally eat the terminator token. It would however still be possible for the node layer to initiate nested lexical spans. This allows the system to model something like `/[/]/` (a regex matching `/`) where existence of an inner lexical span allows the node layer to consume what would otherwise have been a lexical span terminator.

**Comment by conartist6** [comment]
OK I've done enough of the coroutine work to have come back around directly to the issue of how spans will be specified!

**Comment by conartist6** [comment]
Spans are implemented and working in `0.1.12`. Next up: span guards.
